timely_beliefs/beliefs/classes.py

```diff
@@ -487,6 +504,17 @@ def apply_belief_timing_filters(q):
             sources: list = [source] if not isinstance(source, list) else source
             q = q.join(source_class).filter(cls.source_id.in_([s.id for s in sources]))
 
+        if most_recent_belief_only:
+            # most recent belief only makes sense on one defined event
```

Comment:
Actually, it doesn't seem immediately clear to me why we couldn't cover this use case with a combination of `most_recent_beliefs_only=True` and `most_recent_event_only=True`.

Reply:
You still get multiple events if you don't limit to one source.

But foremost, it's really quite slow to use a subquery or even two subqueries. (also with our recent improvement of how the subqueries work)

Reply:
Here I was not suggesting to just use the two old arguments, but to use only 1 of the new arguments, which already limits the query to a single event (and actually, also already limits it to a single belief). It's not obvious to me why the other new argument is needed.

Reply:
Ah, I didn't see you used a new one here.

Two reasons come to mind:
- old argument uses a subquery which is slow
- I added the `most_recent_belief_only` argument also for consistency in arguments.

Reply:
> old argument uses a subquery which is slow

We are free to internally disconnect that logic and switch over to the fast track in case of a specific combination of filters, if it results in identical search results.

> I added the `most_recent_belief_only` argument also for consistency in arguments.

That would make sense, but it doesn't really do what its name suggests. My thinking about this is still evolving. Since my previous comment I realized that picking the most recent event but not the most recent belief about that event doesn't seem like a practical use case.

Current situation:

- The `most_recent_belief_only` filter takes the most recent event and returns the most recent belief about that event that it finds. -> A practical use case.
- The `most_recent_event_only` filter takes the most recent event and returns the first belief about that event that it finds. -> Not a practical use case imo.

My preferred situation involves switching to introducing only one new filter, and renaming it:

- The `most_recent_only` filter takes the most recent event and returns the most recent belief about that event that it finds.

Here I'm disregarding any thoughts on `n_most_recent_events` and probabilistic beliefs.

Reply:
So basically your current thoughts are to

- only add `most_recent_only` filter
- internally switch over to its logic when the user selects one source together with `most_recent_events_only=True`
- document the above point, but also the current state of non-deterministic beliefs being poorly supported here.

Correct?

Reply:
Yes, with one correction:

- *internally switch over to its logic when the user selects one source together with `most_recent_events_only=True`* **and `most_recent_beliefs_only=True`**.

In conjunction, these filters should yield a single row (but only for a sensor holding only deterministic beliefs).

Reply:
Is there a simple & fast way to tell if a sensor has only deterministic beliefs? Otherwise it's not a good idea to do this, maybe.

Reply:
One way would be to count the unique `cumulative_probability` values. But since that would also be new functionality, I agree it is not a good idea to do this within this PR.